tough/src/schema/mod.rs

```diff
@@ -324,6 +392,215 @@ impl Role for Targets {
+            Self::Paths(paths) => {
+                for path in paths {
+                    if Self::matched_path(path, target) {
+                        return true;
+                    }
+                }
+            }
+
+            Self::PathHashPrefixes(path_prefixes) => {
+                for path in path_prefixes {
+                    if Self::matched_prefix(path, target) {
+                        return true;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    ///Given a path hash prefix and a target path determines if target is delegated by prefix
+    fn matched_prefix(prefix: &str, target: &str) -> bool {
+        let temp_target = target.to_string();
+        let hash = digest(&SHA256, temp_target.as_bytes());
+        hash.as_ref().starts_with(prefix.as_bytes())
+    }
+
+    ///Given a shell style wildcard path determines if target matches the path
+    fn matched_path(wildcardpath: &str, target: &str) -> bool {
+        let mut regex_string = wildcardpath.to_string();
+        regex_string = regex_string.replace(".", "\\.");
+        regex_string = regex_string.replace("*", "[^/]*");
+        regex_string = regex_string.replace("?", ".");
+        let re = Regex::new(&regex_string);
+        if let Ok(re) = re {
+            re.is_match(&target)
+        } else {
+            false
+        }
+    }
+}
+
+impl Delegations {
+    ///Determines if target passes pathset specific matching
+    pub fn check_target(&self, target: &str) -> bool {
+        for role in &self.roles {
+            if role.paths.matched_target(target) {
+                return true;
+            }
+        }
+        false
+    }
+
+    ///Ensures that all delegated paths are allowed to be delegated
+    pub fn verify_paths(&self) -> Result<()> {
+        for sub_role in &self.roles {
+            for path in match &sub_role.paths {
+                PathSet::Paths(paths) => paths,
+                PathSet::PathHashPrefixes(paths) => paths,
+            } {
+                if !self.check_target(&path) {
+                    return Err(Error::UnmatchedPath {
+                        child: path.to_string(),
+                    });
+                }
+            }
+        }
+        Ok(())
+    }
+
+    ///Returns given role if its a child of struct
+    pub fn role(&self, role_name: &str) -> Option<&DelegatedRole> {
+        for role in &self.roles {
+            if role.name == role_name {
+                return Some(&role);
+            }
+        }
+        None
+    }
+
+    ///verifies that roles matches contain valid keys
+    pub fn verify_role(&self, role: &Signed<Targets>, name: &str) -> Result<()> {
+        let role_keys = self.role(name).expect("Role not found");
+        let mut valid = 0;
+
+        let mut data = Vec::new();
+        let mut ser = serde_json::Serializer::with_formatter(&mut data, CanonicalFormatter::new());
+        role.signed
+            .serialize(&mut ser)
```

Comment:
I'm a bit confused. Why are we serializing this data?

Reply:
Now that you say that, I'm not sure. I thought it was to make sure it was loaded properly, but serde wouldn't allow it to load wrong anyway.

Reply:
Clarify -- to verify the key ID, which is a digest of the canonical JSON form of the key representation?